File: aoc-rust/all/src/bin/eleven.rs
extern crate all;

use std::collections::HashSet;
use std::collections::hash_map::DefaultHasher;
use std::hash::{Hash, Hasher};

type Floor = u32;

trait ItemClass {
    fn generator(self) -> Option<Item>;
    fn item(self) -> Option<Item>;
}

#[derive(Copy, Clone, Hash, Eq, PartialEq, Debug)]
enum Item {
    TG, TM,
    PLG, PLM,
    PRG, PRM,
    SG, SM,
    RG, RM,
    EG, EM,
    DG, DM
}

impl ItemClass for Item {
    fn generator(self) -> Option<Item> {
        match self {
            Item::TM => Some(Item::TG),
            Item::PLM => Some(Item::PLG),
            Item::PRM => Some(Item::PRG),
            Item::PRM => Some(Item::PRG),
            Item::SM => Some(Item::SG),
            Item::RM => Some(Item::RG),
            Item::DM => Some(Item::DG),
            _ => None
        }
    }

    fn item(self) -> Option<Item> {
        match self {
            Item::TG => Some(Item::TM),
            Item::PLG => Some(Item::PLM),
            Item::PRG => Some(Item::PRM),
            Item::PRG => Some(Item::PRM),
            Item::SG => Some(Item::SM),
            Item::RG => Some(Item::RM),
            Item::DG => Some(Item::DM),
            _ => None
        }
    }
}

#[derive(Clone, PartialEq, Eq, Debug)]
struct State {
    elevator: Floor,
    floors: Vec<HashSet<Item>>
}

impl Hash for State {
    fn hash<H: Hasher>(&self, state: &mut H) {
        state.write_u32(self.elevator);
        let mut statehash = 0u64;
        for floor in &self.floors {
            for item in floor {
                let mut hasher = DefaultHasher::new();
                item.hash(&mut hasher);
                statehash ^= hasher.finish();
            }
        }
        state.write_u64(statehash);
    }
}

impl State {
    fn items_on_floor(&self) -> &HashSet<Item> {
        return &self.floors.get(self.elevator as usize).unwrap();
    }

    fn move_items_to_floor(&self, new_floor: Floor, items: &HashSet<Item>) -> Option<State> {
        let mut new_floors: Vec<HashSet<Item>> = vec![];
        for i in 0..self.floors.len() {
            let floor_items: &HashSet<Item> = self.floors.get(i).unwrap();

            let new_items: HashSet<Item> = if i as u32 == self.elevator {
                floor_items.difference(items).map(Item::clone).collect()
            } else if i as u32 == new_floor {
                floor_items.union(items).map(Item::clone).collect()
            } else {
                floor_items.clone()
            };
//            println!("{} ({}) floor_items: {:?}", self.elevator, i, floor_items);
//            println!("{} ({}) items: {:?}", self.elevator, i, items);
//            println!("{} ({}) new_items: {:?}", self.elevator, i, new_items);

            if !legal_state(&new_items) {
                return None
            }

            new_floors.push(new_items)
        }

        Some(
            State {
                elevator: new_floor,
                floors: new_floors
            }
        )
    }

    fn cost_to_top(&self) -> u32 {
        self.floors.iter().rev().enumerate()
            .fold(0, |acc, (idx, floor)| acc + (idx*floor.len()*1010) as u32) / 2
    }

    fn move_from(&self) -> Vec<State> {
        let mut states = Vec::new();
        if self.elevator > 0 {
            self.new_states_at_floor(&mut states, self.elevator - 1)
        }
        if self.elevator < 3 {
            self.new_states_at_floor(&mut states, self.elevator + 1)
        }

        states
    }

    fn new_states_at_floor(&self, new_states: &mut Vec<State>, floor: Floor) {
        for item_combo in permutations(self.items_on_floor()) {
            if let Some(new_state) = self.move_items_to_floor(floor, &item_combo) {
                new_states.push(new_state);
            }
        }
    }

}

fn permutations(items: &HashSet<Item>) -> Vec<HashSet<Item>> {
    let mut new_items = Vec::new();

    let mut combos_used = HashSet::new();
    for i1 in items {
        let mut s1 = HashSet::new();
        s1.insert(i1.clone());
        new_items.push(s1);
        for i2 in items {
            let new_combo = (i1, i2);
            if i1 != i2 && !combos_used.contains(&new_combo) {
                let mut s2 = HashSet::new();
                s2.insert(i1.clone());
                s2.insert(i2.clone());
                new_items.push(s2);
                combos_used.insert(new_combo);
                combos_used.insert((i2, i1));
            }
        }
    }

    new_items
}

fn legal_state(floor: &HashSet<Item>) -> bool {
    let generators: HashSet<&Item> = floor.iter()
        .filter(|item| item.generator().is_none())
        .collect();

    if !generators.is_empty() {
        return floor
            .iter()
            .filter(|item| !item.generator().is_none())
            .all(|item| generators.contains(&item.generator().unwrap()));
    }

    true
}

/**
public static double costToTop(State state) {
        double cost = 0;

        List<ImmutableSet<Item>> floors = Lists.reverse(state.floors);

        // Cost should be 0 at the top and multiplied by floor at the bottom
        for(int i = 0; i < floors.size(); i++) {
            cost += floors.get(i).size()*i;
        }

        return cost / 2; // up to two items per move
    }
    */


fn main() {
    use all::aoc::astar;
    /*
            State part1 = new State(0, ImmutableList.of(
                ImmutableSet.of(Item.TG, Item.TM, Item.PLG, Item.SG),
                ImmutableSet.of(Item.PLM, Item.SM),
                ImmutableSet.of(Item.PRG, Item.PRM, Item.RG, Item.RM),
                ImmutableSet.of()
            ));
    */
//    let f1: HashSet<Item> = vec![Item::TG, Item::TM, Item::PLG, Item::SG].into_iter().collect();
//    let f2: HashSet<Item> = vec![Item::PLM, Item::SM].into_iter().collect();
//    let f3: HashSet<Item> = vec![Item::PRG, Item::PRM, Item::RG, Item::RM].into_iter().collect();
//    let f4: HashSet<Item> = vec![].into_iter().collect();
//    let part1 = State {
//        elevator: 0,
//        floors: vec![f1, f2, f3, f4]
//    };
    let f1: HashSet<Item> = vec![Item::TG, Item::TM, Item::PLG, Item::SG, Item::EG, Item::EM, Item::DG, Item::DM].into_iter().collect();
    let f2: HashSet<Item> = vec![Item::PLM, Item::SM].into_iter().collect();
    let f3: HashSet<Item> = vec![Item::PRG, Item::PRM, Item::RG, Item::RM].into_iter().collect();
    let f4: HashSet<Item> = vec![].into_iter().collect();
    let part1 = State {
        elevator: 0,
        floors: vec![f1, f2, f3, f4]
    };
//    let f1: HashSet<Item> = vec![Item::RG, Item::TG, Item::TM].into_iter().collect();
//    let f2: HashSet<Item> = vec![].into_iter().collect();
//    let f3: HashSet<Item> = vec![Item::RM].into_iter().collect();
//    let f4: HashSet<Item> = vec![].into_iter().collect();
//    let part1 = State {
//        elevator: 0,
//        floors: vec![f1, f2, f3, f4]
//    };


    let result: astar::AStarResult<State> = astar::search(&part1, State::cost_to_top, State::move_from);
    println!("Test: {:?}", result);
    println!("Test: {:?}", result.path.len() - 1);
}

File: 2017/src/main/java/aoc2016/Eleven.java
package aoc2016;

import aoc.Util;
import com.google.common.collect.ImmutableList;
import com.google.common.collect.ImmutableSet;
import com.google.common.collect.Lists;
import com.google.common.collect.Sets;
import lombok.Value;

import java.util.ArrayList;
import java.util.Collection;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

/**
 * 2016-12-27
 *
 * @author Michael Rose (xorlev)
 */
public class Eleven {
    private static Set<Integer> LEGAL_FLOORS = ImmutableSet.of(0, 1, 2, 3);
    enum Item {
        TG, TM,
        PLG, PLM,
        PRG, PRM,
        SG, SM,
        RG, RM,
        EG, EM,
        DG, DM
    }

    @Value
    public static class State {
        int elevator;
        ImmutableList<ImmutableSet<Item>> floors;

        public State moveItemsToFloor(int newFloor, Set<Item> items) {
            ImmutableList.Builder<ImmutableSet<Item>> newFloors = ImmutableList.builder();
            for(int i = 0; i < floors.size(); i++) {
                Set<Item> floorItems = floors.get(i);

                if(i == elevator) {
                    floorItems = Sets.difference(floorItems, items);
                } else if (i == newFloor) {
                    floorItems = Sets.union(floorItems, items);
                }

                if(!legalState(floorItems))
                    return null;

                newFloors.add(ImmutableSet.copyOf(floorItems));
            }

            return new State(
                newFloor,
                newFloors.build()
            );
        }

        public ImmutableSet<Item> itemsOnFloor() {
            return floors.get(elevator);
        }
    }

    public static Collection<ImmutableSet<Item>> permutations(Set<Item> items) {
        Set<ImmutableSet<Item>> permutations = new HashSet<>();
        for(Item i1 : items) {
            // Add base case: one item
            permutations.add(ImmutableSet.of(i1));
            for(Item i2 : items) {
                permutations.add(ImmutableSet.of(i1, i2));
            }
        }

        return permutations;
    }

    public static List<State> moveFrom(State state) {
        // for floor above and below
        // can we move any combination of items to that floor?

        List<State> states = new ArrayList<>();
        for(Integer floor : Sets.intersection(LEGAL_FLOORS, ImmutableSet.of(state.elevator-1, state.elevator+1))) {
            // all combinations of size <= 2

            for(ImmutableSet<Item> itemCombo : permutations(state.itemsOnFloor())) {
                State newState = state.moveItemsToFloor(floor, itemCombo);

                if(newState != null)
                    states.add(newState);
            }
        }

        return states;
    }

    public static boolean legalState(Collection<Item> floor) {
        Set<Item> generators = new HashSet<>(7);

        for(Item i : floor)
            if(i.name().endsWith("G"))
                generators.add(i);

        // if there is any generator, chips must be paired with their generator
        if(!generators.isEmpty()) {
            for(Item i : floor) {
                if(i.name().endsWith("M")) {
                    // we must have its generator
                    if(!generators.contains(generatorFor(i))) {
                        return false;
                    }
                }
            }
        }

        return true;
    }

    public static Item generatorFor(Item i) {
        return Item.valueOf(i.name().replace('M', 'G'));
    }

    public static double costToTop(State state) {
        double cost = 0;

        List<ImmutableSet<Item>> floors = Lists.reverse(state.floors);

        // Cost should be 0 at the top and multiplied by floor at the bottom
        for(int i = 0; i < floors.size(); i++) {
            cost += floors.get(i).size()*i*101;
        }

        return cost / 2; // up to two items per move
    }

    public static void main(String[] args) throws Exception {
        State easy = new State(0, ImmutableList.of(
            ImmutableSet.of(Item.RG),
            ImmutableSet.of(),
            ImmutableSet.of(Item.RM),
            ImmutableSet.of()
        ));

        /**
         * The first floor contains a thulium generator, a thulium-compatible microchip, a plutonium generator, and a strontium generator.
         The second floor contains a plutonium-compatible microchip and a strontium-compatible microchip.
         The third floor contains a promethium generator, a promethium-compatible microchip, a ruthenium generator, and a ruthenium-compatible microchip.
         The fourth floor contains nothing relevant.
         */
        State part1 = new State(0, ImmutableList.of(
            ImmutableSet.of(Item.TG, Item.TM, Item.PLG, Item.SG),
            ImmutableSet.of(Item.PLM, Item.SM),
            ImmutableSet.of(Item.PRG, Item.PRM, Item.RG, Item.RM),
            ImmutableSet.of()
        ));
        State part2 = new State(0, ImmutableList.of(
            ImmutableSet.of(Item.TG, Item.TM, Item.PLG, Item.SG, Item.EG, Item.EM, Item.DG, Item.DM),
            ImmutableSet.of(Item.PLM, Item.SM),
            ImmutableSet.of(Item.PRG, Item.PRM, Item.RG, Item.RM),
            ImmutableSet.of()
        ));

        // legal floors:
        //  - no generators unless chip's RTG

        // legal moves:
        //  - at least one item
        //  - if two items, it must either be an item & corresponding RTG, or 2 microchips to a floor w/ both RTGs

        Util.AStarResult<State> result1 = Util.timeIt(() -> Util.astarSearch(part1, Eleven::costToTop, Eleven::moveFrom));
        System.out.println(result1);
        System.out.println(result1.getPath().size() - 1);
        System.out.println("---- PART 2 ----");
        Util.AStarResult<State> result2 = Util.timeIt(() -> Util.astarSearch(part2, Eleven::costToTop, Eleven::moveFrom));
        System.out.println(result2);
        System.out.println(result2.getPath().size() - 1);
    }
}
